File: tests/e2e/agent-runtime-golden-path.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { AgentRuntimeFixture } from "../../packages/agent-runtime/src/index.mjs";

const context = {
  schema_version: "1.0", tenant_id: "ten_pet_store", actor_id: "session_customer",
  actor_type: "customer_session", roles: ["customer"], scopes: ["product:read", "agent:run"],
  trace_id: "4bf92f3577b34da6a3ce929d0e0e4736",
  request_origin: { kind: "customer_site", request_id: "req_customer_1" }
};
const product = { tenant_id: "ten_pet_store", product_id: "product_food", name: "Gentle Bites", price_minor: 1299 };

test("customer entry creates correlated run and tenant-scoped typed lookup", () => {
  const runtime = new AgentRuntimeFixture({ products: [product] });
  const result = runtime.lookupProduct({ context, productId: product.product_id, idempotencyKey: "lookup-product-food-1" });
  assert.equal(result.response.success, true);
  assert.equal(result.response.data.product.product_id, "product_food");
  assert.equal(result.response.trace_id, result.run.trace_id);
  assert.equal(runtime.executions.get(result.response.tool_execution_id).agent_run_id, result.run.agent_run_id);
});

test("foreign products are redacted as not found and hostile tenant input fails closed", () => {
  const runtime = new AgentRuntimeFixture({ products: [{ ...product, tenant_id: "ten_other" }] });
  const hidden = runtime.lookupProduct({ context, productId: product.product_id, idempotencyKey: "lookup-foreign-product-1" });
  assert.equal(hidden.response.error.code, "TOOL_NOT_FOUND");
  const started = runtime.startRun(context);
  const hostile = runtime.executeTool({
    context, run: started.run, toolName: "product.lookup",
    input: { product_id: product.product_id, tenant_id: "ten_other" },
    idempotencyKey: "lookup-hostile-input-1"
  });
  assert.equal(hostile.error.code, "TOOL_INVALID_REQUEST");
});

test("approval-gated execution pauses and resumes without changing run identity", () => {
  const runtime = new AgentRuntimeFixture({ products: [product], approve: ({ approvalId }) => approvalId === "approval_1" });
  const started = runtime.startRun(context);
  const paused = runtime.executeTool({
    context, run: started.run, toolName: "product.lookup", input: { product_id: product.product_id },
    idempotencyKey: "approval-product-lookup-1", highRisk: true
  });
  assert.equal(paused.error.code, "TOOL_APPROVAL_REQUIRED");
  assert.equal(started.run.status, "waiting_approval");
  const resumed = runtime.executeTool({
    context, run: started.run, toolName: "product.lookup", input: { product_id: product.product_id },
    idempotencyKey: "approval-product-lookup-1", highRisk: true, approvalId: "approval_1"
  });
  assert.equal(resumed.success, true);
  assert.equal(started.run.agent_run_id, runtime.executions.get(resumed.tool_execution_id).agent_run_id);
});

test("retryable provider failure replays idempotently and conflicts on changed input", () => {
  let calls = 0;
  const runtime = new AgentRuntimeFixture({
    products: [product],
    productProvider: () => { calls += 1; if (calls === 1) throw new Error("provider down"); return product; }
  });
  const first = runtime.lookupProduct({ context, productId: product.product_id, idempotencyKey: "retry-product-lookup-1" });
  assert.equal(first.response.error.code, "TOOL_PROVIDER_FAILED");
  const retry = runtime.resumeRetry({ context, runId: first.run.agent_run_id, productId: product.product_id, idempotencyKey: "retry-product-lookup-1" });
  assert.equal(retry.response.success, true);
  assert.equal(calls, 2);
  const conflict = runtime.executeTool({
    context, run: runtime.runs.get(first.run.agent_run_id), toolName: "product.lookup",
    input: { product_id: "product_other" }, idempotencyKey: "retry-product-lookup-1"
  });
  assert.equal(conflict.error.code, "TOOL_CONFLICT");
});

File: tests/e2e/agent_runtime_golden_path_test.go
package e2e_test

import (
	"errors"
	"testing"

	agentruntime "github.com/yuchuang12/business/packages/agent-runtime"
)

var customerContext = agentruntime.TenantContext{
	SchemaVersion: agentruntime.ContractVersion, TenantID: "ten_pet_store", ActorID: "session_customer",
	ActorType: "customer_session", Roles: []string{"customer"}, Scopes: []string{"product:read", "agent:run"},
	TraceID:       "4bf92f3577b34da6a3ce929d0e0e4736",
	RequestOrigin: agentruntime.RequestOrigin{Kind: "customer_site", RequestID: "req_customer_1"},
}

var fixtureProduct = agentruntime.Product{
	TenantID: "ten_pet_store", ProductID: "product_food", Name: "Gentle Bites", PriceMinor: 1299,
}

func TestCustomerEntryCreatesCorrelatedTenantScopedLookup(t *testing.T) {
	runtime := agentruntime.NewAgentRuntimeFixture([]agentruntime.Product{fixtureProduct}, nil, nil)
	result, err := runtime.LookupProduct(customerContext, fixtureProduct.ProductID, "lookup-product-food-1")
	if err != nil || !result.Response.Success {
		t.Fatalf("lookup = %+v, %v", result, err)
	}
	product := result.Response.Data.(map[string]any)["product"].(agentruntime.Product)
	if product.ProductID != fixtureProduct.ProductID || result.Response.TraceID != result.Run.TraceID {
		t.Fatalf("unexpected response: %+v", result.Response)
	}
	execution, err := runtime.Execution(customerContext, result.Response.ToolExecutionID)
	if err != nil || execution.AgentRunID != result.Run.AgentRunID {
		t.Fatalf("execution = %+v, %v", execution, err)
	}
}

func TestForeignAndReservedTenantInputFailClosed(t *testing.T) {
	runtime := agentruntime.NewAgentRuntimeFixture([]agentruntime.Product{{TenantID: "ten_other", ProductID: fixtureProduct.ProductID}}, nil, nil)
	hidden, err := runtime.LookupProduct(customerContext, fixtureProduct.ProductID, "lookup-foreign-product-1")
	if err != nil || hidden.Response.Error.Code != "TOOL_NOT_FOUND" {
		t.Fatalf("foreign product = %+v, %v", hidden.Response, err)
	}
	started, err := runtime.StartRun(customerContext)
	if err != nil {
		t.Fatal(err)
	}
	hostile, err := runtime.ExecuteTool(agentruntime.ExecuteToolRequest{
		Context: customerContext, RunID: started.Run.AgentRunID, ToolName: "product.lookup",
		Input:          map[string]any{"product_id": fixtureProduct.ProductID, "tenant_id": "ten_other"},
		IdempotencyKey: "lookup-hostile-input-1",
	})
	if err != nil || hostile.Error.Code != "TOOL_INVALID_REQUEST" {
		t.Fatalf("hostile input = %+v, %v", hostile, err)
	}
}

func TestApprovalRetryIdempotencyAndConflict(t *testing.T) {
	approvalRuntime := agentruntime.NewAgentRuntimeFixture([]agentruntime.Product{fixtureProduct}, func(request agentruntime.FixtureApprovalRequest) bool {
		return request.ApprovalID == "approval_1"
	}, nil)
	started, err := approvalRuntime.StartRun(customerContext)
	if err != nil {
		t.Fatal(err)
	}
	paused, err := approvalRuntime.ExecuteTool(agentruntime.ExecuteToolRequest{
		Context: customerContext, RunID: started.Run.AgentRunID, ToolName: "product.lookup",
		Input: map[string]any{"product_id": fixtureProduct.ProductID}, IdempotencyKey: "approval-product-lookup-1", HighRisk: true,
	})
	if err != nil || paused.Error.Code != "TOOL_APPROVAL_REQUIRED" {
		t.Fatalf("pause = %+v, %v", paused, err)
	}
	run, err := approvalRuntime.Service.Run(customerContext, started.Run.AgentRunID)
	if err != nil || run.Status != "waiting_approval" {
		t.Fatalf("paused run = %+v, %v", run, err)
	}
	resumed, err := approvalRuntime.ExecuteTool(agentruntime.ExecuteToolRequest{
		Context: customerContext, RunID: started.Run.AgentRunID, ToolName: "product.lookup",
		Input: map[string]any{"product_id": fixtureProduct.ProductID}, IdempotencyKey: "approval-product-lookup-1", HighRisk: true, ApprovalID: "approval_1",
	})
	if err != nil || !resumed.Success {
		t.Fatalf("resume = %+v, %v", resumed, err)
	}
	execution, err := approvalRuntime.Execution(customerContext, resumed.ToolExecutionID)
	if err != nil || execution.AgentRunID != started.Run.AgentRunID {
		t.Fatalf("resumed execution = %+v, %v", execution, err)
	}

	calls := 0
	retryRuntime := agentruntime.NewAgentRuntimeFixture([]agentruntime.Product{fixtureProduct}, nil, func(product agentruntime.Product) (agentruntime.Product, error) {
		calls++
		if calls == 1 {
			return agentruntime.Product{}, errors.New("provider down")
		}
		return product, nil
	})
	first, err := retryRuntime.LookupProduct(customerContext, fixtureProduct.ProductID, "retry-product-lookup-1")
	if err != nil || first.Response.Error.Code != "TOOL_PROVIDER_FAILED" {
		t.Fatalf("first provider call = %+v, %v", first.Response, err)
	}
	retry, err := retryRuntime.ResumeRetry(customerContext, first.Run.AgentRunID, fixtureProduct.ProductID, "retry-product-lookup-1")
	if err != nil || !retry.Response.Success || calls != 2 {
		t.Fatalf("retry = %+v, calls=%d, err=%v", retry, calls, err)
	}
	conflict, err := retryRuntime.ExecuteTool(agentruntime.ExecuteToolRequest{
		Context: customerContext, RunID: first.Run.AgentRunID, ToolName: "product.lookup",
		Input: map[string]any{"product_id": "product_other"}, IdempotencyKey: "retry-product-lookup-1",
	})
	if err != nil || conflict.Error.Code != "TOOL_CONFLICT" {
		t.Fatalf("conflict = %+v, %v", conflict, err)
	}
}
